rectools/models/base.py

```diff
@@ -230,63 +249,65 @@ def recommend_to_items(
+            target_items, dataset.item_id_map, dataset.n_hot_items, assume_external_ids
         )
+        self._check_targets_are_valid(hot_target_ids, warm_target_ids, cold_target_ids, "item")
 
-        if filter_itself:
-            df_reco = (
-                pd.DataFrame({"tid": reco_target_ids, "iid": reco_item_ids, "score": reco_scores})
-                .query("tid != iid")
-                .groupby("tid", sort=False)
-                .head(k)
-            )
-            reco_target_ids, reco_item_ids, reco_scores = df_reco[["tid", "iid", "score"]].values.T
+        requested_k = k + 1 if filter_itself else k
 
-        if return_external_ids:
-            reco_target_ids = dataset.item_id_map.convert_to_external(reco_target_ids)
-            reco_item_ids = dataset.item_id_map.convert_to_external(reco_item_ids)
+        reco_hot = self._init_internal_reco_triplet()
+        reco_warm = self._init_internal_reco_triplet()
+        reco_cold = self._init_semi_internal_reco_triplet()
+
+        if hot_target_ids.size > 0:
+            reco_hot = self._recommend_i2i(hot_target_ids, dataset, requested_k, sorted_item_ids_to_recommend)
+        if warm_target_ids.size > 0:
+            if self.recommends_for_warm:
+                reco_warm = self._recommend_i2i_warm(
+                    warm_target_ids, dataset, requested_k, sorted_item_ids_to_recommend
+                )
+            else:
+                # TODO: use correct types for numpy arrays and stop ignoring
+                reco_warm = self._recommend_cold(
+                    warm_target_ids, requested_k, sorted_item_ids_to_recommend
+                )  # type: ignore
+        if cold_target_ids.size > 0:
+            reco_cold = self._recommend_cold(cold_target_ids, requested_k, sorted_item_ids_to_recommend)
+
+        reco_hot = self._adjust_reco_types(reco_hot)
+        reco_warm = self._adjust_reco_types(reco_warm)
+        reco_cold = self._adjust_reco_types(reco_cold, target_type=dataset.item_id_map.external_dtype)
 
-        reco = self._make_reco_table(reco_target_ids, reco_item_ids, reco_scores, Columns.TargetItem, add_rank_col)
-        return reco
+        if assume_external_ids:
+            # We have to do it for cold reco before filtering since we need targets and items to be in the same space.
+            # We do it for cold reco only, since it's faster to filter internal ids than external ones.
+            reco_cold_final = self._reco_items_to_external(reco_cold, dataset.item_id_map)
+        else:
+            reco_cold_final = reco_cold
+        del reco_cold
 
-    def _recommend_u2i(
-        self,
-        user_ids: np.ndarray,
-        dataset: Dataset,
-        k: int,
-        filter_viewed: bool,
-        sorted_item_ids_to_recommend: tp.Optional[np.ndarray],
-    ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
-        raise NotImplementedError()
+        if filter_itself:
```

Comment:
It's faster and more memory effective to do this with internal ids

Reply:
> So there is absolutely no way that a cold item could be recommended to itself.

Theoretically there may be an exception from this rule. E.g. if you ask Popular model for 10 recos for cold items and you give it 10 items in `items_to_recommend` list, then it can return you all of them, even if they are cold. However we doesn't support it now.

I agree with the rest. Simplified the code

Reply:
Also can we still drop pandas for filtering?
For example, `_filter_item_itself_from_i2i_reco` will just do the filtering with numpy. And we can make the `head(k)` part in `_make_reco_table` with provided `check_k` argument.

Reply:
But in this case we will execute `.groupby().head()` for all reco, even for cold where we don't need it